projects/pgai/pgai/cli.py

```diff
@@ -218,42 +226,62 @@ def vectorizer_worker(
     poll_interval: int,
     once: bool,
 ) -> None:
+    # gracefully handle being asked to shut down
+    signal.signal(signal.SIGINT, shutdown_handler)
+    signal.signal(signal.SIGTERM, shutdown_handler)
+
     structlog.configure(
         wrapper_class=structlog.make_filtering_bound_logger(get_log_level(log_level))
     )
     log.debug("starting vectorizer worker")
     poll_interval_str = datetime.timedelta(seconds=poll_interval)
 
-    with (
-        psycopg.Connection.connect(db_url) as con,
-        con.cursor(row_factory=namedtuple_row) as cur,
-    ):
-        pgai_version = get_pgai_version(cur)
-        if pgai_version is None:
-            log.critical("the pgai extension is not installed")
-            sys.exit(1)
     dynamic_mode = len(vectorizer_ids) == 0
     valid_vectorizer_ids = []
-    if not dynamic_mode:
-        valid_vectorizer_ids = get_vectorizer_ids(db_url, vectorizer_ids)
-        if len(valid_vectorizer_ids) != len(vectorizer_ids):
-            log.critical(
-                f"invalid vectorizers, wanted: {list(vectorizer_ids)}, got: {valid_vectorizer_ids}"  # noqa: E501 (line too long)
-            )
-            sys.exit(1)
+
+    can_connect = False
+    pgai_version = None
 
     while True:
-        if dynamic_mode:
-            valid_vectorizer_ids = get_vectorizer_ids(db_url, vectorizer_ids)
-            if len(valid_vectorizer_ids) == 0:
-                log.warning("no vectorizers found")
-
-        for vectorizer_id in valid_vectorizer_ids:
-            vectorizer = get_vectorizer(db_url, vectorizer_id)
-            if vectorizer is None:
-                continue
-            log.info("running vectorizer", vectorizer_id=vectorizer_id)
-            run_vectorizer(db_url, vectorizer, concurrency)
+        try:
+            if not can_connect or pgai_version is None:
+                with (
+                    psycopg.Connection.connect(db_url) as con,
+                    con.cursor(row_factory=namedtuple_row) as cur,
+                ):
+                    pgai_version = get_pgai_version(cur)
+                    can_connect = True
+                    if pgai_version is None:
+                        log.warn("the pgai extension is not installed")
+
+            if can_connect and pgai_version is not None:
+                if not dynamic_mode and len(valid_vectorizer_ids) != len(
+                    vectorizer_ids
+                ):
+                    valid_vectorizer_ids = get_vectorizer_ids(db_url, vectorizer_ids)
+                    if len(valid_vectorizer_ids) != len(vectorizer_ids):
+                        log.critical(
+                            f"invalid vectorizers, wanted: {list(vectorizer_ids)}, got: {valid_vectorizer_ids}"  # noqa: E501 (line too long)
+                        )
+                        sys.exit(1)
```

Comment:
Do we want to fail if we request 5 vectorizers and only find 4 (for example)? Why not warn about the one and work the 4?

Reply:
Note: I didn't change that behaviour in this PR.

The original reasoning behind the current behaviour was: "if you've told the vectorizer worker to process a specific set of vectorizers, then it should fail if that specific set isn't completely available". The idea being that it would help in catching misconfigurations.

Reply:
I've changed this now. I've added an `--exit-on-error` parameter. When you specify `--once` it implies `--exit-on-error`. This means that if you run the vectorizer worker in `--once` mode, and tell it you want it to process a specific vectorizer that doesn't exist, it will fail and exit with exit code 1. If you run it in "normal" mode, then it will log a warning, but not exit, unless you tell it to (with `--exit-on-error`).